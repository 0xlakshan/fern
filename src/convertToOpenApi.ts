import { ErrorDeclaration } from "@fern-fern/ir-model/errors";
import { IntermediateRepresentation } from "@fern-fern/ir-model/ir";
import { DeclaredTypeName, TypeDeclaration } from "@fern-fern/ir-model/types";
import { OpenAPIV3 } from "openapi-types";
import { convertServices } from "./converters/servicesConverter";
import { convertType } from "./converters/typeConverter";
import { constructEndpointSecurity, constructSecuritySchemes } from "./security";

export function convertToOpenApi({
    apiName,
    ir,
}: {
    apiName: string;
    ir: IntermediateRepresentation;
}): OpenAPIV3.Document | undefined {
    const schemas: Record<string, OpenAPIV3.ReferenceObject | OpenAPIV3.SchemaObject> = {};

    const typesByName: Record<string, TypeDeclaration> = {};
    ir.types.forEach((typeDeclaration) => {
        // convert type to open api schema
        const convertedType = convertType(typeDeclaration, ir);
        schemas[convertedType.schemaName] = convertedType.openApiSchema;
        // populates typesByName map
        typesByName[getDeclaredTypeNameKey(typeDeclaration.name)] = typeDeclaration;
    });

    const errorsByName: Record<string, ErrorDeclaration> = {};
    ir.errors.forEach((errorDeclaration) => {
        errorsByName[getDeclaredTypeNameKey(errorDeclaration.name)] = errorDeclaration;
    });

    const security = constructEndpointSecurity(ir.auth);

    const paths = convertServices({
        httpServices: ir.services.http,
        typesByName,
        errorsByName,
        errorDiscriminationStrategy: ir.errorDiscriminationStrategy,
        security,
    });
    return {
        openapi: "3.0.1",
        info: {
            title: ir.apiDisplayName ?? apiName,
            description: ir.apiDocs ?? undefined,
            version: "",
        },
        paths,
        components: {
            schemas,
            securitySchemes: constructSecuritySchemes(ir.auth),
        },
        servers: ir.environments.map((environment) => {
            return {
                url: environment.url,
                description:
                    environment.docs != null
                        ? `${environment.name.safeName.originalValue} (${environment.docs})`
                        : environment.name.safeName.originalValue,
            };
        }),
    };
}

export function getDeclaredTypeNameKey(declaredTypeName: DeclaredTypeName): string {
    return `${declaredTypeName.fernFilepath}-${declaredTypeName.name}`;
}
