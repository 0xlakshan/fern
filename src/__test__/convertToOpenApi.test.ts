import { generateIntermediateRepresentation } from "@fern-api/ir-generator";
import { loadWorkspace } from "@fern-api/workspace-loader";
import path from "path";
import { convertToOpenApi } from "../convertToOpenApi";

const FIXTURES = ["test-api", "all-auth", "any-auth"];

describe("convertToOpenApi", () => {
    for (const fixture of FIXTURES) {
        // eslint-disable-next-line jest/valid-title
        it(fixture, async () => {
            const testApiDir = path.join(__dirname, "fixtures", fixture);
            const maybeLoadedWorkspace = await loadWorkspace({
                absolutePathToWorkspace: testApiDir,
                version: 2,
            });
            if (!maybeLoadedWorkspace.didSucceed) {
                throw new Error(JSON.stringify(maybeLoadedWorkspace.failures));
            }
            const intermediateRepresentation = await generateIntermediateRepresentation(maybeLoadedWorkspace.workspace);
            const openApi = convertToOpenApi({
                apiName: "BlogPost API",
                apiVersion: "0.0.0",
                ir: intermediateRepresentation,
            });
            expect(openApi).toMatchSnapshot();
        });
    }
});
