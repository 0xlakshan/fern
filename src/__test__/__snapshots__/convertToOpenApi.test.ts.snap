// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`convertToOpenApi any-auth 1`] = `
"openapi: 3.0.1
info:
  title: fern
  version: ''
paths:
  /create:
    post:
      operationId: PostsService.createPost
      tags:
        - PostsService
      parameters: []
      responses:
        '200':
          description: ''
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/BlogPostId'
      security:
        - BearerAuth: []
        - BasicAuth: []
        - MyHeaderAuth: []
      requestBody:
        required: true
        content:
          application/json:
            schema:
              $ref: '#/components/schemas/BlogBlogPost'
  /{postId}:
    get:
      operationId: PostsService.getPost
      tags:
        - PostsService
      parameters:
        - name: postId
          in: path
          required: true
          schema:
            $ref: '#/components/schemas/BlogPostId'
      responses:
        '200':
          description: ''
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/BlogBlogPost'
components:
  schemas:
    BlogPostId:
      type: string
    BlogAuthor:
      type: object
      properties:
        name:
          type: string
      required:
        - name
    BlogBlogPost:
      type: object
      properties:
        id:
          $ref: '#/components/schemas/BlogPostId'
        title:
          type: string
        author:
          $ref: '#/components/schemas/BlogAuthor'
        content:
          type: string
      required:
        - id
        - title
        - author
        - content
  securitySchemes:
    BearerAuth:
      type: http
      scheme: bearer
    BasicAuth:
      type: http
      scheme: basic
    MyHeaderAuth:
      type: apiKey
      in: header
      name: My-Header
servers: []
"
`;

exports[`convertToOpenApi test-api 1`] = `
"openapi: 3.0.1
info:
  title: fern
  version: ''
paths:
  /create:
    post:
      description: Creates a blog post and returns blog post id.
      operationId: PostsService.createPost
      tags:
        - PostsService
      parameters:
        - name: x-endpoint-header
          in: header
          description: I'm a header!
          required: true
          schema:
            type: string
          examples:
            Example1:
              value: abc
      responses:
        '200':
          description: ''
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/BlogPostId'
              examples:
                Example1:
                  value: post-id-123
      requestBody:
        required: true
        content:
          application/json:
            schema:
              $ref: '#/components/schemas/BlogCreatePostRequest'
            examples:
              Example1:
                value:
                  title: My Blog
                  author:
                    type: anonymous
                  content: I'm a post
                  postType: LONG
  /{postId}:
    get:
      description: Gets a blog post.
      operationId: PostsService.getPost
      tags:
        - PostsService
      parameters:
        - name: postId
          in: path
          required: true
          schema:
            $ref: '#/components/schemas/BlogPostId'
          examples:
            Example1:
              value: post-id-123
        - name: some-query-param
          in: query
          required: true
          schema:
            type: integer
          examples:
            Example1:
              value: 123
      responses:
        '200':
          description: ''
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/BlogBlogPost'
        '400':
          description: ''
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/BlogPostNotFoundErrorBody'
              examples:
                Example1:
                  value:
                    id: post-id-123
    post:
      description: Updates a blog post.
      operationId: PostsService.updatePost
      tags:
        - PostsService
      parameters:
        - name: postId
          in: path
          required: true
          schema:
            $ref: '#/components/schemas/BlogPostId'
      responses:
        '200':
          description: ''
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/BlogPostId'
        '400':
          description: ''
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/BlogPostNotFoundErrorBody'
      requestBody:
        required: true
        content:
          application/json:
            schema:
              $ref: '#/components/schemas/BlogCreatePostRequest'
components:
  schemas:
    BlogPostId:
      type: string
    BlogBlogPost:
      type: object
      description: A blog post
      properties:
        id:
          $ref: '#/components/schemas/BlogPostId'
        type:
          $ref: '#/components/schemas/BlogPostType'
        title:
          type: string
        author:
          $ref: '#/components/schemas/BlogAuthor'
        content:
          type: string
      required:
        - id
        - type
        - title
        - author
        - content
    BlogPostType:
      type: string
      enum:
        - LONG
        - SHORT
    BlogAuthor:
      oneOf:
        - type: object
          properties:
            type: {}
        - type: object
          properties:
            type:
              type: string
        - type: object
          allOf:
            - $ref: '#/components/schemas/BlogPostType'
            - type: object
              properties:
                type:
                  type: string
                  enum:
                    - type
    BlogPostNotFoundErrorBody:
      type: object
      properties:
        id:
          $ref: '#/components/schemas/BlogPostId'
      required:
        - id
    BlogCreatePostRequest:
      type: object
      properties:
        title:
          type: string
        author:
          $ref: '#/components/schemas/BlogAuthor'
          description: Author is the writer.
        content:
          type: string
        postType:
          $ref: '#/components/schemas/BlogPostType'
      required:
        - title
        - author
        - content
        - postType
  securitySchemes: {}
servers:
  - url: https://buildwithfern.com
  - url: https://dev.buildwithfern.com
"
`;
